spec/formatting.md

```diff
@@ -18,3 +18,71 @@ the local variable takes precedence.
+
+## Error Handling
+
+During the formatting of a message,
+various errors may be encountered.
+These are divided to the following categories:
+
+- **Syntax errors** occur when the syntax representation of a message is invalid.
+- **Resolution errors** occur when the runtime value of a part of a message
+  cannot be determined.
+
+  - **Unresolved Variable** errors occur when a variable reference cannot be resolved.
+
+- **Selection errors** cover failures encountered during selection.
+
+  - **Selector errors** are failures in the matching of a key to a specific selector.
+  - **Missing Fallback** errors occur when no Variant is selected
+    due to the message not including a Variant with only catch-all keys.
+
+- **Formatting errors** occur during the formatting of a resolved value,
+  for example when encountering a value with an unsupported type
+  or an internally inconsistent set of options.
+
+During selection, an expression handler must only emit Resolution and Selection errors.
+During formatting, an expression handler must only emit Resolution and Formatting errors.
+
+In all cases, when encountering an error,
+a message formatter must provide some representation of the message.
+An informative error or errors must also be separately provided.
+
+When an error occurs in the syntax or resolution of an Expression or MarkupStart Option,
+the Expression or MarkupStart in question is processed as if the option was not defined.
+This may allow for the fallback handling described below to be avoided,
+though an error must still be emitted.
+
+When an error occurs within a Selector,
+the selector must not match any VariantKey other than the catch-all `*`
+and a Selector error is emitted.
+When selection fails to match any Variant,
+an empty string is used as the formatted string representation of the message
+and a Missing Fallback error is emitted.
+
+When an error occurs in a Placeholder that is being formatted,
+the fallback string representation of the Placeholder
+always starts with U+007B LEFT CURLY BRACKET `{`
+and ends with U+007D RIGHT CURLY BRACKET `}`.
+Between the brackets, the following contents are used:
+
+- Expression with Literal Operand: U+0028 LEFT PARENTHESIS `(`
+  followed by the value of the Literal,
+  and then by U+0029 RIGHT PARENTHESIS `)`
+- Expression with Variable Operand: U+0024 DOLLAR SIGN `$`
+  followed by the Variable Name of the Operand
+- Expression with no Operand: U+003A COLON `:` followed by the Expression Name
+- Markup start: U+002B PLUS SIGN `+` followed by the MarkupStart Name
+- Markup end: U+002D HYPHEN-MINUS `-` followed by the MarkupEnd Name
+- Otherwise: Three U+003F QUESTION MARK `?` characters, i.e. `???`
+
+For example, the formatted string representation of the expression `{$foo :bar}`
+would be `{$foo}` if the variable could not be resolved.
+
+The formatted string representation of a message with an unrecoverable syntax error
```

Comment:
Can/should we allow a formatter to emit a fatal (such as throwing an exception) on a parse/syntax error (for cases where the formatter separates construction from the call to the format function)? This line of text only defines what the output is for cases where we do *not* emit a fatal before calling format, which I think is necessary.

I was going to suggest that you omit the word "syntax" (i.e. just "unrecoverable error"), but note that we allow the empty string for missing variant keys above. Should those behaviors be identical or should they remain separate?

Reply:
I would prefer being able to treat even syntax errors as non-fatal. If the UI wants _some_ string representation for a message it should get that, even if the message's syntax is broken, or if it's otherwise not available.

For a broken-syntax message, I would think that something like `{msg-id}` or `{���}` would be better than an empty string. If we want to use the same representation here as for the missing-variant case, we should not use an empty string.

Reply:
> I would prefer being able to treat even syntax errors as non-fatal. If the UI wants some string representation for a message it should get that, even if the message's syntax is broken, or if it's otherwise not available.

We are arguing opposite sides of the street :-). I agree that implementations should be ***able*** to treat syntax errors as non-fatal. But we can also allow implementations to optionally "fail loudly" (equivalent of `throws`)

I agree that broken messages can emit the "broken message placeholder"

Reply:
> I would prefer being able to treat even syntax errors as non-fatal.
> If the UI wants some string representation for a message it should get that

Not sure how that would work.
For a malformed message that representation would probably be completely useless.

> I would think that something like {msg-id}

There is no such thing as message id at this level.
In general a message id is used by a resource manager to retrieve a string, which is passed to MF2 to be parsed and so on. So MF2 gets the string (serialized format of the message), but not the ID.
It one wants to produce something like {msg-id}, then it is at a different level, not MF2.

In other words: the recovery for syntax error belongs to the level on top of MF2.
One such level might actually add more info (for example source file & line number).
Which is info not available to MF2.

So syntax errors are fatal, and is the job of the layer above to handle them
(similar to error handling considered good practices for programming in general).

Reply:
The "failing loudly" bit here is now addressed at least in part by this line:
> Syntax and Data Model errors must be emitted as soon as possible.

Effectively, this ought to allow for an implementation which has some kind of a message construction phase (e.g. parsing the message source) separately from a formatting call, to fail differently for syntax or data model errors compared to later runtime errors.

Separately, the proposal has been updated to remove references to any message identifiers, instead referring to a "fallback string" that may be provided externally.